### ISSUE TITLE
Convert filtering functions to classes

### ISSUE THREAD

**Issue by niekdejonge** [matchms/matchms#434]
The filtering steps now are functions with similar code structure. 

It would make sense to create a parent class for the FilterMethod, to formalize what is expected of these filtering functions. Just like we did with the MetadataFeatureGenerator in MS2Deepscore.

Benefit would be cleaner methods and easier to add general new functionality.
A few things that are now already standard are:
input is type Spectrum.
Always run:
if spectrum_in is None:
    return None

spectrum = spectrum_in.clone()

Output is of type Optional[Spectrum]

We could also make more specific Classes. Some ideas:
CleanMetadata which can only access metadata of spectra
CleanPeaks which can only change peaks and intensities
Require which can not change spectra, but only checks if a requirement is met and returns None or the Spectrum. 
This is closely related to #432 

It would make it easier to implement future changes that apply to every filtering function. It would also make it possible to more safely use your own filtering methods in pipeline, since they would require to be of type FilterMethod. 

Downside might be that not all people in the community are very familiar with using classes, with might make it less accessible to contribute. 

What do you think @florian-huber @hechth

**Comment by niekdejonge** [comment]
Also code like in add_parent_mass (and many other filtering functions): 
logger = logging.getLogger("matchms")

_default_key = "parent_mass"
_accepted_keys = ["parentmass", "exact_mass"]
_accepted_types = (float, str, int)
_accepted_missing_entries = ["", "N/A", "NA", "n/a"]

can be included in init.

**Comment by niekdejonge** [comment]
It might also be possible to add a method for required pre filtering for each filter step, that specify which other filter steps are expected to already have been done. This could be used in Pipeline, during the precheck if the order of filtering steps makes sense. For instance you should not run add_parent_mass before derive_ionmode or correct_charge, since it depends on the entry of charge and ionmode.

**Comment by florian-huber** [comment]
As discussed in person, I would prefer to stick to the filter functions. But as outlined in #432, I would rather try adding a `SpectrumProcessor` class that will take care of running those functions. This could include:
- A few default settings, e.g., `minimal`, `default`, `full` processing options (or also a flag for `GCMS` or `LCMS`?)
- Also the option to list all filters that should be applied, the class will then take care of the right order of those filters (e.g., add, repair etc. parent mass before running a filter relying on it).

--> See #455.

**Comment by florian-huber** [comment]
Several points came up again in the making of the `SpectrumProcessor` class (see #472 #455 #470 ). So, I here come back to the question of how to implement the individual filters.

@niekdejonge you suggested using classes, which would be a way do remove code duplications. For instance something like this:
```python
class BaseSpectrumFilter:
    def __init__(self):
        pass

    def filter(self, spectrum_in: SpectrumType) -> SpectrumType:
        if spectrum_in is None:
            return None
        spectrum = spectrum_in.clone()
        return self.apply_filter(spectrum)

    def apply_filter(self, spectrum: SpectrumType) -> SpectrumType:
        """This method should be implemented by child classes."""
        raise NotImplementedError


class XYZFilter(BaseSpectrumFilter):
    def apply_filter(self, spectrum: SpectrumType) -> SpectrumType:
        # Do the specific filtering for XYZ here.
        # ...
        return spectrum
```

I initially was skeptical about this. First, because I am weary of classes with only one method (always seems like they don't need to be classes.... but that's a bit minimalistic). And they would also require two lines to be executed because we always need to first create an instance:

```python
xyz_filter = XYZFilter()
result = xyz_filter.filter(input_spectrum)
```
Both things, however, are no big issues. So, I am not strongly against this solution! 


A possible functional alternative could be to use decorators:
```python
def spectrum_filter_decorator(filter_function):
    def wrapper(spectrum_in: SpectrumType, **args) -> SpectrumType:
        if spectrum_in is None:
            return None
        spectrum = spectrum_in.clone()
        return filter_function(spectrum, **args)
    return wrapper

@spectrum_filter_decorator
def filter_xyz(spectrum: SpectrumType, **args) -> SpectrumType:
    # Do the specific XYZ filtering here.
    # ...
    return spectrum
```
As I said.... not sure myself what's better.
And to make choices even more complicated, here a 3rd option: Combining classes with functions by having wrapper functions like:
```python
def apply_xyz_filter(spectrum_in: SpectrumType) -> SpectrumType:
    return XYZFilter().filter(spectrum_in)
```

**Comment by niekdejonge** [comment]
I agree that SpectrumProcessor and the regrouping of the filters solve the most important points already. 
I did not think of the decorator option before, but that might indeed be a good alternative indeed. 

A few additional use cases I had in mind for the classes solution would be for instance:
- You could specify required_other_filters in a Filter attribute. For instance, derive_ionmode would have make_ionmode_lowercase as a required filter, which can be easily checked by pipeline to see if anything in the pipeline is conflicting. 
- Different filter types, For instance a RequireFilter, which cannot make changes but can only return None or the old spectrum. This would formalize expected behaviour of different filter types, which makes the code clearer, but also might make implementation of (future) architecture of pipeline easier or faster. 

But I do not think that any of these changes are really important, they mostly would feel like an elegant implementation. 
So I agree that this will not be worth the effort for now. 

We could indeed add the decorator to reduce the code duplication. On the other hand, SpectrumProcessor now also already checks if a Spectrum is None and clones spectra. So this might not even be necessary anymore.

**Comment by florian-huber** [comment]
See also comment from @hechth in #455 :

> since some filters share quite a lot of functionality which could be put into a single place this way - specifically about things like logging or exception handling - some of this can also be put into the SpectrumProcessor class but I think we should first actually sketch out how to implement it. All existing functions can also be mapped to the object oriented design and marked as deprecated to not break the current API.

> As of now, with all filters being individual functions, I find it quite hard to harmonize the behaviour, logging messages etc. - these things could then be handled by the base classes and on the highest layer, everything is just implementing the Filter interface which has a single apply function or so, and then the specialized base classes can implement the specific behaviour - like the require filters either return the same object or None - this could then even be checked in a RequireFilter class where apply runs the implementation of the specific filtering method (let's call it require in this case) and then checks whether the spectrum is still the same or None - else it should throw an exception or so. Overall, the Template method pattern can be used here really nicely.

> @zargham-ahmad and me have been sketching out an implementation based on this design choice - maybe we can move this implementation to a specific feature branch for the filters before moving it into the main codebase?

@hechth @niekdejonge Guess I should be convinced by now 😄 
 --> since this seems to be internal refactoring which will not break the main user interface parts, moving the implementation to a filter branch first would be great!

**Comment by florian-huber** [comment]
In the current `development` branch towards `matchms 1.0.0.dev1` it is still filter functions, but now including several factory functions. Happy Class or not question again, but then on the new code which includes both handling `Spectrum` objects as well as `SpectraCollection` objects.
